virtual_rainforest/models/hydrology/below_ground.py

```diff
@@ -195,3 +199,77 @@ def convert_soil_moisture_to_water_potential(
     return air_entry_water_potential * (
         (soil_moisture / soil_moisture_capacity) ** water_retention_curvature
     )
+
+
+def update_groundwater_storge(
+    groundwater_storage: NDArray[np.float32],
+    vertical_flow_to_groundwater: NDArray[np.float32],
+    bypass_flow: NDArray[np.float32],
+    max_percolation_rate_uzlz: Union[float, NDArray[np.float32]],
+    groundwater_loss: Union[float, NDArray[np.float32]],
+    reservoir_const_upper_groundwater: Union[float, NDArray[np.float32]],
+    reservoir_const_lower_groundwater: Union[float, NDArray[np.float32]],
+) -> dict[str, NDArray[np.float32]]:
+    """Update groundwater storage and calculate below ground horizontal flow.
+
+    Groundwater storage and transport are modelled using two parallel linear reservoirs,
+    similar to the approach used in the HBV-96 model
+    :cite:p:`lindstrom_development_1997` and the LISFLOOD
+    :cite:p:`van_der_knijff_lisflood_2010`.
+    The upper zone represents a quick runoff component, which includes
+    fast groundwater and subsurface flow through macro-pores in the soil. The lower zone
+    represents the slow groundwater component that generates the base flow.
+
+    Args:
+        groundwater_storage: amount of water that is stored in the groundwater reservoir
+            , [mm]
+        vertical_flow_to_groundwater: flux from the lower soil layer to groundwater for
+            this timestep, [mm]
+        bypass_flow: flow that bypasses the soil matrix and drains directly to the
+            groundwater, [mm]
+        max_percolation_rate_uzlz: maximum perclation rate between pper and lower
+            groundwater zone, [mm d-1]
+        groundwater_loss: constant amount of water that never rejoins the river channel
+            and is lost beyond the catchment boundaries or to deep groundwater systems,
+            [mm]
+        reservoir_const_upper_groundwater: reservoir constant for the upper groundwater
+            layer, [days]
+        reservoir_const_lower_groundwater: reservoir constant for the lower groundwater
+            layer, [days]
+
+    Returns:
+        updated amount of water stored in upper and lower zone, outflow from the upper
+            zone to the channel, and outflow from the lower zone to the channel
+    """
+
+    output = {}
```

Comment:
I'd make a dataclass for the output variable, so that the data structure is a bit more self-documenting. Plus you get better typing that way.

Reply:
Could you please give me an example for this? I have heard of data classes but I don't know how I could implement your suggestion such that the function returns a data class.

Reply:
First define a dataclass for this return type -- something like this:

```py
from dataclasses import dataclass

@dataclass
class GroundwaterStorage:
    subsurface_flow: NDArray[np.float32]
    baseflow: NDArray[np.float32]
    updated_groundwater_storage: NDArray[np.float32]
```

Then you can use it like so:

```py
output = GroundwaterStorage(subsurface_flow=some_value, baseflow=other_value, updated_groundwater_storage=third_value)
```

Dataclasses are like dicts but just let you be explicit about what attributes you want and what their types are.